lib/metasploit/framework/credential_collection.rb

```diff
@@ -1,233 +1,292 @@
-          end
-          if user_as_pass
-            yield Metasploit::Framework::Credential.new(public: user_from_file, private: user_from_file, realm: realm, private_type: :password)
-          end
-          if blank_passwords
-            yield Metasploit::Framework::Credential.new(public: user_from_file, private: "", realm: realm, private_type: :password)
+    # Returns true when #each will have no results to iterate
+    def empty?
+      prepended_creds.empty? && !has_privates?
+    end
+
+    def has_privates?
+      password.present? || pass_file.present? || !additional_privates.empty? || blank_passwords
+    end
+
+    protected
+
+    def private_type(private)
+      if private =~ /[0-9a-f]{32}:[0-9a-f]{32}/
+        :ntlm_hash
+      elsif private =~ /^md5([a-f0-9]{32})$/
+        :postgres_md5
+      else
+        :password
+      end
+    end
+  end
+
+  class CredentialCollection < PrivateCredentialCollection
+
+    # @!attribute additional_publics
+    #   Additional public to be combined
+    #
+    #   @return [Array<String>]
+    attr_accessor :additional_publics
+
+    # @!attribute user_as_pass
+    #   Whether each username should be tried as a password for that user
+    #   @return [Boolean]
+    attr_accessor :user_as_pass
+
+    # @!attribute user_file
+    #   Path to a file containing usernames, one per line
+    #   @return [String]
+    attr_accessor :user_file
+
+    # @!attribute username
+    #   @return [String]
+    attr_accessor :username
+
+    # @!attribute userpass_file
+    #   Path to a file containing usernames and passwords separated by a space,
+    #   one pair per line
+    #   @return [String]
+    attr_accessor :userpass_file
+
+    # @option opts [Boolean] :blank_passwords See {#blank_passwords}
+    # @option opts [String] :pass_file See {#pass_file}
+    # @option opts [String] :password See {#password}
+    # @option opts [Array<Credential>] :prepended_creds ([]) See {#prepended_creds}
+    # @option opts [Boolean] :user_as_pass See {#user_as_pass}
+    # @option opts [String] :user_file See {#user_file}
+    # @option opts [String] :username See {#username}
+    # @option opts [String] :userpass_file See {#userpass_file}
+    def initialize(opts = {})
+      super
+      self.additional_publics  ||= []
+    end
+
+    # Adds a string as an addition public credential
+    # to be combined in the collection.
+    #
+    # @param [String] public_str the string to use as a public
+    # @return [void]
+    def add_public(public_str='')
+      additional_publics << public_str
+    end
+
+    # Combines all the provided credential sources into a stream of {Credential}
+    # objects, yielding them one at a time
+    #
+    # @yieldparam credential [Metasploit::Framework::Credential]
+    # @return [void]
+    def each
+      if pass_file.present?
+        pass_fd = File.open(pass_file, 'r:binary')
+      end
```

Comment:
Given that his has multiple checks for the presence of a username, I'm not sure how that would work without defining more methods.

Reply:
Fair this is more of a refactoring comment to use more methods rather than reusing code so often but given that a lot of this has to do with the username and we aren't really focused on refactoring at the moment I don't think it makes sense to get hung up on this.